### PR TITLE
fix(Core): Save gameobject state on instances 

### PR DESCRIPTION
**Pull request by claudiodfc**
## Changes Proposed:
Currently, azerothcore doesn't save gameobject states on instances.
Whenever there's a re-start or crash, the instance's gameobjects and their states aren't saved, producing un-wanted behaviours and blocking instances at times.

Implemented CRUD for new table `instance_saved_go_state_data` that holds the states of gameobjects.

- When worldserver launches and gameobjects are loaded, this will check if this object's state exists on the DB and sets the previous state.
- On instance deletion (reset) these states are also removed based on the instance ID.
- Whenever a gameobject state changes inside a dungeon or raid, we save on the database the set state.

## Issues Addressed:

- Closes https://github.com/azerothcore/azerothcore-wotlk/issues/9422
- Closes https://github.com/azerothcore/azerothcore-wotlk/issues/9468
- Closes part of https://github.com/azerothcore/azerothcore-wotlk/issues/9414
- Closes part of https://github.com/azerothcore/azerothcore-wotlk/issues/9529

## SOURCE:

No source

## Tests Performed:

- Tested in-game

## How to Test the Changes:

The following test case should be executed when in a party and solo, to see if the instances are resetting properly.

If possible testing some dungeons like the ones mentioned on the issues would be appreciated.

#### Test case 1
1. .tele culling
2. .go gameobject 67455
3. .gobject activate 67455
4. .ser res 1
5. Re-launch Worldserver
6. Check if book-case is still open
7. Repeat steps 1, 2 and 3 but Reset all instances after repeating the first step.

## Known Issues and TODO List:

- [x] There might be some crash so if possible doing other dungeon test such as the Deadmines would help out. I wasn't able to reproduce a crash again.

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

File: data/sql/updates/pending_db_world/rev_1643395587559675400.sql [hunk 1]
```diff
@@ -0,0 +1,9 @@
+INSERT INTO `version_db_world` (`sql_rev`) VALUES ('1643395587559675400');
+
+CREATE TABLE IF NOT EXISTS `instance_saved_data` (
+  `instance_id` int unsigned NOT NULL,
+  `gameobject_entry` int unsigned NOT NULL,
+  `gameobject_guid` int unsigned NOT NULL,
+  `gameobject_state` int unsigned DEFAULT '0',
+  PRIMARY KEY (`instance_id`,`gameobject_entry`,`gameobject_guid`)
+) ENGINE=InnoDB DEFAULT CHARSET=utf8mb3 COLLATE=utf8_bin;
```
Comment (Kitzunu) [inline data/sql/updates/pending_db_world/rev_1643395587559675400.sql:9]: https://www.azerothcore.org/wiki/sql-standards#tables-and-columns

use utf8mb4, also it is good if you add DROP TABLE IF EXISTS just to follow the way tables are already created

File: src/server/game/Entities/GameObject/GameObject.cpp [hunk 2]
```diff
@@ -2411,6 +2438,79 @@ void GameObject::SetGoState(GOState state)
...
+            saveInstanceData(&state);
+        }
+        else
+        {
+            updateInstanceData(&state);
+        }
+    }
+}
+
+void GameObject::saveInstanceData(GOState* state)
```
Comment (Kitzunu) [inline src/server/game/Entities/GameObject/GameObject.cpp:2459]:
```suggestion
void GameObject::SaveInstanceData(GOState* state)
```

File: src/server/game/Entities/GameObject/GameObject.cpp [hunk 3]
```diff
@@ -2411,6 +2438,79 @@ void GameObject::SetGoState(GOState state)
...
+{
+    CharacterDatabasePreparedStatement* stmt = CharacterDatabase.GetPreparedStatement(CHAR_INSERT_INSTANCE_SAVED_DATA);
+    stmt->setUInt32(0, GetInstanceId());
+    stmt->setUInt32(1, GetEntry());
+    stmt->setUInt32(2, GetSpawnId());
+    stmt->setUInt32(3, *state);
+    CharacterDatabase.Execute(stmt);
+}
+
+void GameObject::updateInstanceData(GOState* state)
```
Comment (Kitzunu) [inline src/server/game/Entities/GameObject/GameObject.cpp:2469]:
```suggestion
void GameObject::UpdateInstanceData(GOState* state)
```

File: src/server/game/Entities/GameObject/GameObject.cpp [hunk 4]
```diff
@@ -2411,6 +2438,79 @@ void GameObject::SetGoState(GOState state)
...
+{
+    CharacterDatabasePreparedStatement* stmt = CharacterDatabase.GetPreparedStatement(CHAR_UPDATE_INSTANCE_SAVED_DATA);
+    stmt->setUInt32(0, *state);
+    stmt->setUInt32(1, GetSpawnId());
+    stmt->setUInt32(2, GetInstanceId());
+    CharacterDatabase.Execute(stmt);
+}
+
+bool GameObject::hasStateSavedOnInstance()
```
Comment (Kitzunu) [inline src/server/game/Entities/GameObject/GameObject.cpp:2478]:
```suggestion
bool GameObject::HasStateSavedOnInstance()
```

File: src/server/game/Entities/GameObject/GameObject.cpp [hunk 5]
```diff
@@ -2411,6 +2438,79 @@ void GameObject::SetGoState(GOState state)
...
+    {
+        // There's no gameobject with this GUID saved on the DB
+        return false;
+    }
+
+    // Found gameobject on the database saved
+    return true;
+}
+
+int8 GameObject::getInstanceSavedState()
```
Comment (Kitzunu) [inline src/server/game/Entities/GameObject/GameObject.cpp:2496]:
```suggestion
int8 GameObject::GetInstanceSavedState()
```

File: src/server/game/Entities/Player/PlayerMisc.cpp [hunk 6]
```diff
@@ -181,6 +181,16 @@ void Player::SendResetFailedNotify(uint32 mapid)
     GetSession()->SendPacket(&data);
 }
 
+void deleteInstanceSavedData(uint32 instanceId)
```
Comment (Kitzunu) [inline src/server/game/Entities/Player/PlayerMisc.cpp:184]:
```suggestion
void DeleteInstanceSavedData(uint32 instanceId)
```

File: src/server/game/Groups/Group.cpp [hunk 7]
```diff
@@ -2013,6 +2013,16 @@ void Group::SetRaidDifficulty(Difficulty difficulty)
     }
 }
 
+void Group::resetInstanceSavedGameobjects(uint32 instanceId)
```
Comment (Kitzunu) [inline src/server/game/Groups/Group.cpp:2016]:
```suggestion
void Group::ResetInstanceSavedGameobjects(uint32 instanceId)
```

**Comment by Kitzunu** [comment]
You will need to correct the capitalization in the other places as well as I only made the suggestion for where the functions were defined.

Otherwise, this looks cool

File: data/sql/updates/pending_db_world/rev_1643395587559675400.sql [hunk 8]
```diff
@@ -0,0 +1,9 @@
+INSERT INTO `version_db_world` (`sql_rev`) VALUES ('1643395587559675400');
+
+CREATE TABLE IF NOT EXISTS `instance_saved_data` (
+  `instance_id` int unsigned NOT NULL,
+  `gameobject_entry` int unsigned NOT NULL,
+  `gameobject_guid` int unsigned NOT NULL,
+  `gameobject_state` int unsigned DEFAULT '0',
```
Comment (Kitzunu) [inline data/sql/updates/pending_db_world/rev_1643395587559675400.sql:7]: I'd like to say id, entry, guid, state

**Comment by Nyeriah** [comment]
I think we need startup sanitization, so we don't up with orphaned rows in case the instances are reset manually through the database or some other weird thing happens

File: data/sql/updates/pending_db_world/rev_1643395587559675400.sql (hunk 8 above)
Comment (Nyeriah) [reply data/sql/updates/pending_db_world/rev_1643395587559675400.sql:7]: that seems better yep, its simpler, we can keep it simple

File: data/sql/updates/pending_db_world/rev_1643395587559675400.sql [hunk 9]
```diff
@@ -2,9 +2,9 @@ INSERT INTO `version_db_world` (`sql_rev`) VALUES ('1643395587559675400');
...
-  `gameobject_guid` int unsigned NOT NULL,
-  `gameobject_state` int unsigned DEFAULT '0',
-  PRIMARY KEY (`instance_id`,`gameobject_entry`,`gameobject_guid`)
-) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4;
+  `id` int unsigned NOT NULL,
+  `entry` int unsigned NOT NULL,
+  `guid` int unsigned NOT NULL,
+  `state` int unsigned DEFAULT '0',
+  PRIMARY KEY (`id`,`entry`,`guid`) USING BTREE
+) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COMMENT='id = instance.id\r\nentry = gameobject.entry\r\nguid = gameobject.guid\r\nstate = gameobject''s state';
```
Comment (Kitzunu) [inline data/sql/updates/pending_db_world/rev_1643395587559675400.sql:10]: I'd suggest adding these comments to the column fields instead

**Comment by Winfidonarleyan** [comment]
maybe `instance_saved_data` rename to `instance_saved_go_state_data` or other?

File: src/server/game/Entities/GameObject/GameObject.cpp [hunk 10]
```diff
@@ -2411,6 +2438,79 @@ void GameObject::SetGoState(GOState state)
...
+    {
+        // There's no gameobject with this GUID saved on the DB
+        return 0;
+    }
+
+    Field* fields = result->Fetch();
+
+    // Found gameobject on the database saved
+    return fields[0].GetInt8();
 }
```
Comment (Winfidonarleyan) [inline src/server/game/Entities/GameObject/GameObject.cpp:2514]: this bad functions. This is a heavy load on the database

**Comment by claudiodfc** [comment]
So how do I make them better?

**Comment by claudiodfc** [comment]
By container you mean loading the data from DB onto the memory? Maybe async select is better?

**Review by Nyeriah — APPROVED** [review]

**Comment by Kitzunu** [comment]
dont care about codefactor

File: data/sql/updates/pending_db_world/rev_1643395587559675400.sql [hunk 11]
```diff
@@ -0,0 +1,10 @@
+INSERT INTO `version_db_world` (`sql_rev`) VALUES ('1643395587559675400');
+
+DROP TABLE IF EXISTS `instance_saved_go_state_data`;
+CREATE TABLE IF NOT EXISTS `instance_saved_go_state_data` (
+  `id` int unsigned NOT NULL COMMENT 'id = instance.id',
+  `entry` int unsigned NOT NULL COMMENT 'entry = gameobject.entry',
+  `guid` int unsigned NOT NULL COMMENT 'guid = gameobject.guid',
+  `state` int unsigned DEFAULT '0' COMMENT 'state = gameobject.state',
```
Comment (Kitzunu) [inline data/sql/updates/pending_db_world/rev_1643395587559675400.sql:8]:
```suggestion
  `id` INT UNSIGNED NOT NULL COMMENT 'instance.id',
  `entry` INT UNSIGNED NOT NULL COMMENT 'gameobject_template.entry',
  `guid` INT UNSIGNED NOT NULL COMMENT 'gameobject.guid',
  `state` INT UNSIGNED DEFAULT '0' COMMENT 'gameobject.state',
```

File: src/server/game/Globals/ObjectMgr.cpp [hunk 12]
```diff
@@ -9352,6 +9352,33 @@ void ObjectMgr::LoadFactionChangeQuests()
     LOG_INFO("server.loading", " ");
 }
 
+void ObjectMgr::LoadInstanceSavedGameobjectStateData()
```
Comment (Kitzunu) [inline src/server/game/Globals/ObjectMgr.cpp:9355]: should probably move this to where gameobjects are loaded or somewhere better and not mid FactionChange loads :D

File: data/sql/updates/pending_db_world/rev_1643395587559675400.sql (hunk 11 above)
Comment (Kitzunu) [reply data/sql/updates/pending_db_world/rev_1643395587559675400.sql:8]: Actually, when I am thinking about it. Why do we store entry? Does it really serve any purpose?

File: src/server/game/Entities/GameObject/GameObject.h [hunk 13]
```diff
@@ -982,6 +982,20 @@ class GameObject : public WorldObject, public GridObject<GameObject>, public Mov
...
+    {
+        if (auto* map = FindMap())
+        {
+            if (map->IsDungeon() || map->IsRaid())
+            {
+                return true;
+            }
+        }
+        return false;
+    };
```
Comment (Winfidonarleyan) [inline src/server/game/Entities/GameObject/GameObject.h:998]: I think it is better to do the implementation in cpp

File: src/server/game/Entities/Player/Player.h [hunk 14]
```diff
@@ -2407,7 +2407,6 @@ class Player : public Unit, public GridObject<Player>
     void SetLastPetNumber(uint32 petnumber) { m_lastpetnumber = petnumber; }
     [[nodiscard]] uint32 GetLastPetSpell() const { return m_oldpetspell; }
     void SetLastPetSpell(uint32 petspell) { m_oldpetspell = petspell; }
-
```
Comment (Winfidonarleyan) [inline src/server/game/Entities/Player/Player.h:2410]: no need del this line

File: src/server/game/Globals/ObjectMgr.h [hunk 15]
```diff
@@ -1374,6 +1374,33 @@ class ObjectMgr
...
+        {
+            if (it->m_entry == entry && it->m_guid == guid && it->m_instance == id)
+            {
+                it->m_state = state;
+            }
+        }
+    }
+    void NewInstanceSavedGameobjectState(uint32 id, uint32 entry, uint32 guid, uint8 state)
+    {
+        GameobjectInstanceSavedStateList.push_back({ id, entry, guid, state });
```
Comment (Winfidonarleyan) [inline src/server/game/Globals/ObjectMgr.h:1402]:
```suggestion
        GameobjectInstanceSavedStateList.emplace_back(id, entry, guid, state);
```

File: src/server/game/Globals/ObjectMgr.h [hunk 16]
```diff
@@ -1374,6 +1374,33 @@ class ObjectMgr
...
+    void SetInstanceSavedGameobjectState(uint32 id, uint32 entry, uint32 guid, uint8 state)
+    {
+        for (auto it = GameobjectInstanceSavedStateList.begin(); it != GameobjectInstanceSavedStateList.end(); it++)
+        {
+            if (it->m_entry == entry && it->m_guid == guid && it->m_instance == id)
+            {
+                it->m_state = state;
+            }
+        }
+    }
```
Comment (Winfidonarleyan) [inline src/server/game/Globals/ObjectMgr.h:1399]: I think it is better to do the implementation in cpp

**Comment by claudiodfc** [comment]
@Nyeriah why is the build failling?

```
In mysql_stmt_prepare() id: 448, sql: "UPDATE instance_saved_go_state_data SET state = ? WHERE guid = ? AND id = ?"
Table 'acore_characters.instance_saved_go_state_data' doesn't exist
In mysql_stmt_prepare() id: 449, sql: "INSERT INTO instance_saved_go_state_data (id, entry, guid, state) VALUES (?, ?, ?, ?)"
Table 'acore_characters.instance_saved_go_state_data' doesn't exist
In mysql_stmt_prepare() id: 450, sql: "DELETE FROM instance_saved_go_state_data WHERE id = ?"
Table 'acore_characters.instance_saved_go_state_data' doesn't exist
In mysql_stmt_prepare() id: 451, sql: "DELETE FROM instance_saved_go_state_data WHERE id NOT IN (SELECT instance.id FROM instance)"
Table 'acore_characters.instance_saved_go_state_data' doesn't exist
Could not prepare statements of the Character database, see log for details.
```

**Comment by Nyeriah** [comment]
> @Nyeriah why is the build failling?
> 
> ```
> In mysql_stmt_prepare() id: 448, sql: "UPDATE instance_saved_go_state_data SET state = ? WHERE guid = ? AND id = ?"
> Table 'acore_characters.instance_saved_go_state_data' doesn't exist
> In mysql_stmt_prepare() id: 449, sql: "INSERT INTO instance_saved_go_state_data (id, entry, guid, state) VALUES (?, ?, ?, ?)"
> Table 'acore_characters.instance_saved_go_state_data' doesn't exist
> In mysql_stmt_prepare() id: 450, sql: "DELETE FROM instance_saved_go_state_data WHERE id = ?"
> Table 'acore_characters.instance_saved_go_state_data' doesn't exist
> In mysql_stmt_prepare() id: 451, sql: "DELETE FROM instance_saved_go_state_data WHERE id NOT IN (SELECT instance.id FROM instance)"
> Table 'acore_characters.instance_saved_go_state_data' doesn't exist
> Could not prepare statements of the Character database, see log for details.
> ```

You've added the SQL to /pending_world/ instead of /pending_characters/, so the table is being created in the `world` database while the core expects it to be in `characters`

**Comment by claudiodfc** [comment]
Loool force of habit

**Comment by Kitzunu** [comment]
> Actually, when I am thinking about it. Why do we store entry? Does it really serve any purpose?

**Comment by claudiodfc** [comment]
I implemented in for future purposes when there's hooks for gameobjects with the same entry. It doesn't hurt.

**Comment by Nyeriah** [comment]
It actually does because that's a lot of data :P but if you sure it'll be useful maybe we could look past that problem

**Comment by claudiodfc** [comment]
Alright, give me a few days and I'll take care of it, it's not needed for now and it's not so mandatory

**Comment by Kitzunu** [comment]
```
/home/runner/work/azerothcore-wotlk/azerothcore-wotlk/src/server/game/Globals/ObjectMgr.cpp: In member function ‘void ObjectMgr::LoadInstanceSavedGameobjectStateData()’:
/home/runner/work/azerothcore-wotlk/azerothcore-wotlk/src/server/game/Globals/ObjectMgr.cpp:9646:145: error: no matching function for call to ‘std::vector<ObjectMgr::GameobjectInstanceSavedState>::push_back(<brace-enclosed initializer list>)’
         GameobjectInstanceSavedStateList.push_back({ fields[0].GetUInt32(), fields[1].GetUInt32(), fields[2].GetUInt32(), fields[3].GetUInt8() });

**Comment by claudiodfc** [comment]
Thank you for the suggestion Nyeriah, I though about this too and I've got no rush to merge this, as long as you guys can cope with the amount of commits I'm doing to this

**Comment by claudiodfc** [comment]
@temperrr Could you give this a shot whenever you can? I'm needing a bit of feedback on possible issues, otherwise It's pretty good to go I would say.

**Comment by claudiodfc** [comment]
##### Test case 1

- Activated a gameobject (Img.1)
- Relogged to save character near that gameobject
- Restarted Server
- Gameobject state is loaded as previous state (Img.2)

##### Img 1
image


##### Img 2
image

File: src/server/game/Entities/GameObject/GameObject.cpp [hunk 17]
```diff
@@ -2411,6 +2434,122 @@ void GameObject::SetGoState(GOState state)
...
+        case GAMEOBJECT_TYPE_TRAPDOOR:
+            return true;
+        default:
+            return false;
+    }
+}
+
+unsigned short GameObject::GetGameobjectStateAsShort(GOState* state)
+{
+    unsigned short m_state = 3;
```
Comment (Winfidonarleyan) [inline src/server/game/Entities/GameObject/GameObject.cpp:2491]: uint16

**Comment by claudiodfc** [comment]
why uint16 for such a small number?

File: src/server/game/Entities/GameObject/GameObject.cpp (hunk 17 above)
Comment (Kitzunu) [reply src/server/game/Entities/GameObject/GameObject.cpp:2491]: or uint8, just not unsigned short

Comment (Nyeriah) [reply src/server/game/Entities/GameObject/GameObject.cpp:2491]: should match the gameobject *state* attribute imo

File: data/sql/updates/pending_db_characters/rev_1643629468629316100.sql [hunk 18]
```diff
@@ -0,0 +1,9 @@
+INSERT INTO `version_db_characters` (`sql_rev`) VALUES ('1643629468629316100');
+
+DROP TABLE IF EXISTS `instance_saved_go_state_data`;
+CREATE TABLE IF NOT EXISTS `instance_saved_go_state_data` (
+  `id` INT UNSIGNED NOT NULL COMMENT 'instance.id',
+  `guid` INT UNSIGNED NOT NULL COMMENT 'gameobject.guid',
+  `state` INT UNSIGNED DEFAULT '0' COMMENT 'gameobject.state',
```
Comment (Kitzunu) [inline data/sql/updates/pending_db_characters/rev_1643629468629316100.sql:7]:
```suggestion
  `state` TINYINT UNSIGNED DEFAULT '0' COMMENT 'gameobject.state',
```

File: src/server/game/Entities/GameObject/GameObject.cpp (hunk 17 above)
Comment (Kitzunu) [reply src/server/game/Entities/GameObject/GameObject.cpp:2491]: Yea uint8, the DB table should change to TINYINT

Comment (Kitzunu) [reply src/server/game/Entities/GameObject/GameObject.cpp:2491]: And this func should be renamed as well

Comment (Kitzunu) [reply src/server/game/Entities/GameObject/GameObject.cpp:2491]: (state in `gameobject` is TINYINT)

**Comment by claudiodfc** [comment]
@azerothcore/testers need some feedback on this, let me know if you need help testing

**Comment by temperrr** [comment]
image

**Comment by Lynethris** [comment]
Screenshot 2022-02-26 212730

**Comment by devovh** [comment]
Fix: https://github.com/devovh/azerothcore-wotlk/commit/175b35ce7d6e9a7ca48a4187823e0de555a73abb

**Comment by Nyeriah** [comment]
Why was this closed?

**Comment by acidmanifesto** [comment]
> Why was this closed?

It was never able to be compiled to begin and the PR maker was the only person able to provide screen shots of a test.

**Comment by Nyeriah** [comment]
> > Why was this closed?
> 
> It was never able to be compiled to begin and the PR maker was the only person able to provide screen shots of a test.

This PR was working, however, the database rewrite took place inbetween the tests so it needed to be updated accordingly. It just needs the queries updated so it can be tested again and hopefully merged.

**Comment by claudiodfc** [comment]
Was going to fix the mess you made with my PR, but without any warning, notice and change-log, by merging un-tested and shit code that has no "Why are we changing something" type of description, I honestly don't care and can't bother to help anymore.
Don't forget to keep copy pasting code from Trinitycore and not developing your core as a team, administrators.

**Comment by Nyeriah** [comment]
Hey @claudiodfc 

I understand you're frustrated with the changes that eventually hit your PR while you were working on it and we also have acknowledged that the PR that introduced them lacked the change log entry and the BREAKING CHANGE tag, however, the implementation itself comes as a vast improvement to our database layer, as we no longer need to specify the data types when executing queries.

For example, where otherwise you'd see:
```cpp
query->SetUint32(0, my_variable);
```

Now it simply resolves to:

```cpp
query->SetData(0, my_variable);
```

As exemplified in the PR description. This allows for a much simpler use of it and reduces the code complexity, making it easier to use and lowering the overall necessary knowledge to write code.

I understand that it is frustrating that with every major rewrite several new issues show up and things need to be fixed again, but overall I believe it is for the best. We are moving forward, even though the journey might seem often bumpy at times.

If you find yourself in trouble with any of the changes you're free to throw me a DM or ask in one of the AC channels, I'm certain people will be more than happy to lend you a hand, because, afterall, we're all here united by the same goal: make things better for everyone.

Regards

**Comment by FrancescoBorzi** [comment]
>  and not developing your core as a team, administrators

@claudiodfc if there is a problem feel free to approach me via DM. BTW, I can no longer find you on Discord.

**Pull request closed without merging**